### ISSUE TITLE
Disk Cache: Add a Sensible Default for Key Encoder

### ISSUE THREAD

**Issue by AndrewBarba** [kean/Nuke#160]
In the current example for `enableExperimentalAggressiveDiskCaching` you are using a sha1 hash as the filename based on the images url, can you think of any reason we can't just use `String.hashValue`? Asking because it's just more accessible in my opinion

```swift
ImagePipeline.shared = ImagePipeline {
    $0.enableExperimentalAggressiveDiskCaching { String($0.hashValue) }
}
```

**Comment by kean** [comment]
Hey @AndrewBarba. Yes, you do need to use one of the _cryptographic_ hash functions which guarantee that the probability of two URLs having the same hash value (which is used as a filename) is so low that you can ignore it. `sha1` is a good candidate.

The `hashValue` is a different kind of hash function which is designed to be extremely fast (much faster than cryptographic hash functions) but doesn't give many guarantees in terms of collisions. These types of hash functions are designed to be used primarily by hash tables in which keys are eventually tested for equality. If you were to use `hashValue` as a filename you are guaranteed to encounter a situation where you produce the same filename for two different URLs.

Another alternative that I've explored is percent encoding URLs to produce valid file names. The problem with that approach is that filenames have size limits. The maximum file name size on HFS+ is 255 characters and I'm not yet sure what the limit is on APFS.

**Comment by kean** [comment]
There is another option, if you know that you can extract some ID from a URL which uniquely identifies an image _and_ can be used as a filename (or you can percent encode it if necessary), you can use it as a filename.

**Comment by AndrewBarba** [comment]
I thought about doing that but most of the images I'm using are going through a service like Imgix which has sizing params in the url so the id doesn't do me much good. Thanks for the explanation though! Will most likely stick with sha1 for now

**Comment by kean** [comment]
Yep, this is perfect.

**Comment by AndrewBarba** [comment]
I don’t have hard numbers for you, but the app I work on (recently updated to nuke 7, disk cache with sha256) has been scrolling with no hiccups at all. In the main table view there are 3 images loaded by nuke per cell and it fits 4 cells on screen. The perceived performance has been excellent.

**Comment by kean** [comment]
I'm not really expecting any differences in perceived performance, but I prefer to use the fastest stuff available anyway 😃 The compile time is a true concern though. I wouldn't want users to measure compile time at some point and find out that there are some Nuke's functions at the top (700 ms for a single function is a lot).

**Comment by AndrewBarba** [comment]
Yes totally get it. And that’s why I use Nuke over everything else :)

**Comment by kean** [comment]
Swift 4.2 finally seems to come with `CommonCrypto` https://twitter.com/mxcl/status/1003789764884140038. If you wait long enough...

**Comment by kean** [comment]
I've implemented a default `FilenameGenerator` which uses SHA1. It's going to ship in v7.3 and is going to be available in Swift 4.2.
